pkg/process/util/containers.go

```diff
@@ -6,104 +6,280 @@
-	for _, name := range names {
-		detectors = append(detectors, collectors.NewDetector(name))
+// GetSharedContainerProvider returns a shared ContainerProvider
+func GetSharedContainerProvider() ContainerProvider {
+	initContainerProvider.Do(func() {
+		sharedContainerProvider = NewDefaultContainerProvider()
+	})
+	return sharedContainerProvider
+}
+
+// NewContainerProvider returns a ContainerProvider instance
+func NewContainerProvider(provider metrics.Provider, metadataStore workloadmeta.Store, filter *containers.Filter) ContainerProvider {
+	return ContainerProvider{
+		metricsProvider: provider,
+		metadataStore:   metadataStore,
+		filter:          filter,
 	}
-	dedupe = len(detectors) > 1
+}
+
+// NewDefaultContainerProvider returns a ContainerProvider built with default metrics provider and metadata provider
+func NewDefaultContainerProvider() ContainerProvider {
+	containerFilter, err := containers.GetSharedMetricFilter()
+	if err != nil {
+		log.Warnf("Can't get container include/exclude filter, no filtering will be applied: %w", err)
+	}
+
+	return NewContainerProvider(metrics.GetProvider(), workloadmeta.GetGlobalStore(), containerFilter)
 }
 
 // GetContainers returns containers found on the machine
-// GetContainers autodetects the best backend from available sources
-// if the users don't specify the preferred container sources
-func GetContainers() ([]*containers.Container, error) {
-	// Detect sources
-	if detectors == nil {
-		// Container sources aren't configured, autodetect the best available source
-		detectors = []*collectors.Detector{collectors.NewDetector("")}
-	}
-
-	result := []*containers.Container{}
-	for _, detector := range detectors {
-		l, name, err := detector.GetPreferred()
+func (p *ContainerProvider) GetContainers(cacheValidity time.Duration, previousContainers map[string]*ContainerRateMetrics, previousTime, currentTime time.Time) ([]*model.Container, map[string]*ContainerRateMetrics, map[int]string, error) {
+	containersMetadata, err := p.metadataStore.ListContainers()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	hostCPUCount := float64(system.HostCPUCount())
+	processContainers := make([]*model.Container, 0)
+	rateStats := make(map[string]*ContainerRateMetrics)
+	pidToCid := make(map[int]string)
+	for _, container := range containersMetadata {
+		if !container.State.Running {
+			continue
+		}
+
+		if p.filter != nil && p.filter.IsExcluded(container.Name, container.Image.Name, container.Labels["io.kubernetes.pod.namespace"]) {
+			continue
+		}
+
+		if container.Runtime == workloadmeta.ContainerRuntimeGarden && len(container.CollectorTags) == 0 {
+			log.Debugf("No tags found for garden container: %s, skipping", container.ID)
+			continue
+		}
+
+		entityID := containers.BuildTaggerEntityName(container.ID)
+		tags, err := tagger.Tag(entityID, collectors.HighCardinality)
 		if err != nil {
-			return nil, err
+			log.Warnf("Could not collect tags for container %q, err: %w", container.ID[:12], err)
```

Comment:
This is going to be potentially very noisy.

Reply:
Agreed, downgraded to `DEBUG`.